File: #0096.unique-binary-search-trees.cpp
class Solution {
public:
    int numTrees(int n) {

        vector<int> result(n + 1);
        result[0] = 1;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                result[i] += result[j] * result[i - j - 1];
            }
        }
        return result[n];
    }
};

File: #0096.unique-binary-search-trees.rs
impl Solution {
    pub fn num_trees(n: i32) -> i32 {

        let n = n as usize;
        let mut result = vec![0; n + 1];
        result[0] = 1;
        for i in 1..=n {
            for j in 1..=i {
                result[i] += result[j - 1] * result[i - j];
            }
        }

        return result[n];
    }
}
